daemon/openrazer_daemon/hardware/device_base.py

```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                            if not 0 <= self.zone[i]["colors"][index] <= 255:
+                                raise ValueError('Color out of range')
+
+                        if len(self.zone[i]["colors"]) != 9:
+                            raise ValueError('There must be exactly 9 colors')
+
+                    except ValueError:
+                        # invalid colors. reinitialize
+                        self.zone[i]["colors"] = [0, 255, 0, 0, 255, 255, 0, 0, 255]
+                        self.logger.info("%s: Invalid colors; restoring to defaults.", self.__class__.__name__)
+                        pass
+
+                    except KeyError:
+                        # no colors. reinitialize
+                        self.zone[i]["colors"] = [0, 255, 0, 0, 255, 255, 0, 0, 255]
+                        self.logger.info("%s: No colors found; restoring to defaults.", self.__class__.__name__)
+                        pass
+
+                    try:
+                        self.zone[i]["speed"] = int(self.config[self.storage_name][i + '_speed'])
+                        self.zone[i]["wave_dir"] = int(self.config[self.storage_name][i + '_wave_dir'])
+
+                    except KeyError:
+                        self.zone[i]["speed"] = 1
+                        self.zone[i]["wave_dir"] = 1
+                        self.logger.info("%s: No speed/wave direction found; restoring to defaults.", self.__class__.__name__)
+                        pass
+
+                    if i == "backlight":
+                        effect_func_name = 'set' + self.zone[i]["effect"][0].upper() + self.zone[i]["effect"][1:]
+                    else:
+                        effect_func_name = 'set' + i[0].upper() + i[1:] + self.zone[i]["effect"][0].upper() + self.zone[i]["effect"][1:]
+                else:
+                    self.zone[i]["effect"] = 'spectrum'
+                    effect_func_name = 'setSpectrum'
+                    if i == "backlight":
+                        effect_func_name = 'setSpectrum'
+                    else:
+                        effect_func_name = 'set' + i[0].upper() + i[1:] + 'Spectrum'
+
+                effect_func = getattr(self, effect_func_name, None)
+
+                if effect_func == None and not self.zone[i]["effect"] == "Spectrum":
+                    self.logger.info("%s: Invalid effect name %s; restoring to Spectrum.", self.__class__.__name__, effect_func_name)
+                    self.zone[i]["effect"] = 'spectrum'
+                    if i == "backlight":
+                        effect_func_name = 'setSpectrum'
+                    else:
+                        effect_func_name = 'set' + i[0].upper() + i[1:] + 'Spectrum'
+                    effect_func = getattr(self, effect_func_name, None)
+
+                if not effect_func == None:
+                    if self.get_num_arguments(effect_func) == 1:
+                        if self.zone[i]["effect"] == 'starlightRandom':
```

Comment:
Not doing this one, because you didn't realize one thing: it'd require a lot more code since `effect_func_name` may change per zone (e.g. `setSpectrum` and `setLogoSpectrum`)